lib/cancan/ability/strong_parameter_support.rb

```diff
@@ -0,0 +1,25 @@
+module CanCan
+  module Ability
+    module StrongParameterSupport
+      # Return an array of attributes suitable for use with strong parameters
+      def permitted_attributes(action, subject)
+        @permitted_attributes ||= {}
+        @permitted_attributes[[action, subject]] ||= allowed_attributes(action, subject)
+      end
+
+      private
+
+      def allowed_attributes(action, subject)
+        attributes = relevant_rules(action, subject).flat_map do |rule|
```

Comment:
wouldn't it be more performant to implement it using `Set` and `rule.base_behavior`? So, something like this:

```rb
attributes = relevant_rules(action, subject).inject(Set.new) do |set, rule|
  if rule.attributes.empty? && subject < ActiveRecord::Base
    attrs = subject.column_names.map(&:to_sym) - [:id]
  else 
    attrs = rule.attributes
  end

  if rule.base_behavior
     set.add(attrs)
  else
     set.subtract(attrs)
  end

  set
end
```

Eventually we get unique list of attributes and ensure that user is able to access them

Reply:
Eventually you can use `attributes.to_a` to convert them into an array